TC-CNET-4.11

aabgd opened the pull request:
#### Testing

This PR migrates Test_TC_CNET_4_11.yaml to TC_CNET_4_11.py file. This PR is linked to Fixes [#409](https://github.com/project-chip/matter-test-scripts/issues/409)

The test specification is [4.2.9. TC-CNET-4.11](https://github.com/CHIP-Specifications/chip-test-plans/blob/master/src/cluster/network_commissioning.adoc#429-tc-cnet-411-wi-fi-verification-for-connectnetwork-command-dut-server).
Test Spec updated with [PR #5221](https://github.com/CHIP-Specifications/chip-test-plans/pull/5221)

Updates:

- Created TC_CNET_4_11.py
- Removed Test_TC_CNET_4_11.yaml
- Updated test_metadata.yaml

Testing:
Since this is a wifi test, to reproduce test steps there are some physical devices needed.
You will need:

Device: **ESP32C6**
WiFi: you'll need two WiFi networks with corresponding SSIDs and Passphrases that will be provided in the CLI to run the script.

```
In one terminal:
# 1. check out ESP-IDF tool and install it
#    using the instructions on the page below
# https://project-chip.github.io/connectedhomeip-doc/platforms/esp32/setup_idf_chip.html

Get ESP-IDF v5.3
mkdir ~/esp
cd ~/esp
git clone -b v5.3 --recursive --depth 1 --shallow-submodule https://github.com/espressif/esp-idf.git
cd ~/esp/esp-idf
./install.sh

# 2. go to connectedhomeip project home
source scripts/bootstrap.sh
source scripts/activate.sh
python3 -m pip install esptool 

# 3. Run export.sh 
source ../esp/esp-idf/export.sh

# 4. Run activate.sh again
source scripts/activate.sh

# 5. IDF set-target
idf.py -C examples/all-clusters-app/esp32 esp32c6
# probably you will need to remove examples/all-clusters-app/esp32/build before running previous command
# you will need to modify examples/all-clusters-app/esp32/partitions.csv to re-flash
ota_0,    app,  ota_0,   ,        2500K,
# ota_0,    app,  ota_0,   ,        1900K,
# ota_1,    app,  ota_1,   ,        1900K,

# 6. IDF build
idf.py -C examples/all-clusters-app/esp32 build

Remove flash between runs:
idf.py -C examples/all-clusters-app/esp32 -p /dev/tty.usbserial-140 erase-flash

# 7. IDF flash
idf.py -C examples/all-clusters-app/esp32 -p /dev/tty.usbserial-140 flash monitor
# (in my case on Mac the device is called /dev/tty.tty.usbserial-140)
# (in my case on Linux the device is called /dev/ttyUSB0)
```

In a second terminal:

Run test:
```
python3 src/python_testing/TC_CNET_4_11.py --commissioning-method ble-wifi --wifi-ssid <SSID1> --wifi-passphrase <Passphrase1> --discriminator 3840 --passcode 20202021 --string-arg PIXIT.CNET.WIFI_2ND_ACCESSPOINT_SSID:"SSID2" --string-arg PIXIT.CNET.WIFI_2ND_ACCESSPOINT_CREDENTIALS:"Passphrase2" --endpoint 0
```

#### Fixes to TC_CNET_4_11.py for Docker container compatibility inside TH:

- Remove sudo when running as root — added `privileged_cmd()` helper; containers running as root don't have sudo installed
- Fix infinite hang in `wpa_command()` — added 5s socket timeout on `recvfrom()`; missing reply from `wpa_supplicant` previously froze the entire `asyncio` event loop
- Fix `wpa_supplicant` reply path in Docker — moved client socket bind path from `/tmp` (container-private) to `/var/run/wpa_supplicant/` (bind-mounted from host); `wpa_supplicant` couldn't deliver replies back to the container-local path
- Gracefully skip `dhclient` when absent — added `dhclient_renew()` helper with `shutil.which()` check; NetworkManager handles DHCP internally when `dhclient` binary is not installed

#### Run this test on TH:

Make sure to **bind-mount** the host's `wpa_supplicant` path into the Docker container using the -v flag:: `-v /var/run/wpa_supplicant:/var/run/wpa_supplicant`

e.g.:

```
docker run -v /var/credentials:/credentials \
        -v /var/run/dbus/system_bus_socket:/var/run/dbus/system_bus_socket \
        -v /home/ubuntu/certification-tool/backend/test_collections/matter/sdk_tests/sdk_checkout/python_testing:/root/python_testing \
        -v $(pwd):/launch_dir \
        -v /var/run/wpa_supplicant:/var/run/wpa_supplicant \
        --privileged \
        --network host \
        -it connectedhomeip/chip-cert-bins:e9ecfc2138887d3221dcc2995ad629c8bd4313e4
```

aabgd committed:
started with TC_CNET_4_11.py

aabgd committed:
added step precondition

aabgd committed:
steps precondition 1 and 2 to be reviewed in chromebook

aabgd committed:
trying to connect to wifi

aabgd committed:
updated steps from precondition to step 6

aabgd committed:
cleaned old code

aabgd committed:
added steps 7, 9, 10, 11 and 14

aabgd committed:
added steps 13, 15, 16, 17, 19, 20, 21, 22

aabgd committed:
abstracted away some steps and created functions

aabgd committed:
corrected some errors

aabgd committed:
corrected more errors

aabgd committed:
started refactoring to use matter_testing functions

aabgd committed:
continued refactoring

aabgd committed:
connected host to 2nd wifi network

aabgd committed:
1st pass all tests

aabgd committed:
removed commented code and added timeouts

aabgd committed:
added asyncio.sleep(5) to change_networks

aabgd committed:
moved connect_wifi_host from step 13 up to step 12

aabgd committed:
added retries to connect_host_wifi to avoid network unstability

aabgd committed:
worked on change_networks

aabgd committed:
made functions async, added some async.wait, tried to force CASE session restablishment, fails on step 13 as mentioned in Bug #26848

aabgd committed:
cleaned up code, added test_metadata

aabgd committed:
created connect_wifi_linux and connect_wifi_mac

aabgd committed:
updated connect_wifi_linux

aabgd committed:
changed TIMEOUT to 900

aabgd committed:
removed logs that are not needed

aabgd committed:
restored previous version, added constants and retries 1/50

aabgd committed:
updated Step 13

aabgd committed:
adding read_single_attribute_all_endpoints for reading networks

aabgd committed:
added read_single_attribute_all_endpoints and updated Step 9

aabgd committed:
changed TIMED_REQUEST_TIMEOUT_MS

aabgd committed:
cleaned change_networks and updated MAX_RETRIES

aabgd committed:
abstracted away verify_operational_network and updated steps 9, 13 and 19, changed WIFI_WAIT_SECONDS to 5

aabgd committed:
updated Step 11

aabgd committed:
Removed manual steps and updated steps numbers

aabgd committed:
Removed unnecesary logs

aabgd committed:
cleaned logs and comments, ready for PR

aabgd committed:
removed Test_TC_CNET_4_11.yaml

aabgd committed:
updated according to Gemini comments in PR #39311

aabgd commented on src/python_testing/TC_CNET_4_11.py:
```diff
@@ -0,0 +1,603 @@
+    os_name = platform.system()
+    logger.info(f" --- connect_host_wifi: OS detected: {os_name}")
+
+    # Let's try to connect the TH to the second network a few times, to avoid network instability
```
Added os_name = None

aabgd commented on src/python_testing/TC_CNET_4_11.py:
```diff
@@ -0,0 +1,603 @@
+attr = cnet.Attributes
+
+
+async def connect_wifi_linux(ssid, password):
```
Added brief docstrings

aabgd committed:
added conn = None and removed os_name = None in connect_host_wifi

aabgd committed:
removed [ in logger.error in connect_wifi_mac

aabgd committed:
updated docstrigns for helper functions

aabgd committed:
solved linter issues in PR

aabgd committed:
Merge branch 'master' into tc_cnet_4_11

aabgd committed:
Merge branch 'master' into tc_cnet_4_11

aabgd committed:
Merge branch 'master' into tc_cnet_4_11

aabgd commented on src/python_testing/TC_CNET_4_11.py:
```diff
@@ -0,0 +1,611 @@
+                logger.info(f" --- verify_operational_network: DUT connected to SSID: {ssid}")
+                break
+
+    def steps_TC_CNET_4_11(self):
```
Hi, there's a new [PR 5221](https://github.com/CHIP-Specifications/chip-test-plans/pull/5221) already in place for the test spec as mentioned in the first comment, though will add the expected outcome to the .py script, and remove PICs and Ref columns from the test spec. Thanks.

aabgd replied:
Added expected outcome to the test steps. For the other points Issue 40251 was created.

aabgd committed:
Merge branch 'master' into tc_cnet_4_11

aabgd commented on src/python_testing/TC_CNET_4_11.py:
```diff
@@ -0,0 +1,611 @@
+MAX_RETRIES = 5
+TIMEOUT = 900
+TIMED_REQUEST_TIMEOUT_MS = 5000
+WIFI_WAIT_SECONDS = 5
```
`WIFI_WAIT_SECONDS` is a constant used when changing networks in the TH (not the DUT), something that wasn't present in the framework and mostly used in the helper functions.
`TIMEOUT` is a constant used to set up the test framework's `default_timeout` property and as an argument to `wait_for` calls.
`TIMED_REQUEST_TIMEOUT_MS` is a constant used to pass `timedRequestTimeoutMs` when calling  `self.send_single_cmd`
All of them set up in one place as these are used all over the script.

aabgd committed:
Merge branch 'master' into tc_cnet_4_11

aabgd committed:
updated expected outcomes from steps 1 to 4

aabgd committed:
updated expected outcomes from steps 5 to 10

aabgd committed:
updated expected outcomes from steps 11 to 15

aabgd committed:
updated expected outcomes from steps 16 to 20

khodya commented on src/python_testing/TC_CNET_4_11.py:
```diff
@@ -0,0 +1,611 @@
+        return result
+
+
+async def connect_host_wifi(ssid, password):
```
could you put these function in `chip/testing/wifi.py` file? please do it in a separate PR as a follow up

aabgd replied:
Created [Issue #40426](https://github.com/project-chip/connectedhomeip/issues/40426)

khodya reviewed (approved):
LGTM, given that the open comments will be resolved in the follow up tasks

aabgd committed:
Merge branch 'master' into tc_cnet_4_11

aabgd committed:
Separated the logic of verifying if an ssid is present in networks in the method find_network_and_assert

aabgd commented on src/python_testing/TC_CNET_4_11.py:
```diff
@@ -0,0 +1,611 @@
+            asserts.fail(f" --- verify_operational_network: Could not read networks after {MAX_RETRIES} retries.")
+
+        asserts.assert_is_not_none(networks, "Could not read networks.")
+        for idx, network in enumerate(networks):
```
Done.

aabgd commented on src/python_testing/TC_CNET_4_11.py:
```diff
@@ -0,0 +1,611 @@
+        # 1. NetworkID is the hex representation of the ASCII values for PIXIT.CNET.WIFI_2ND_ACCESSPOINT_SSID
+        # 2. Connected is of type bool and is FALSE
+        userwifi_2nd_netidx = None
+        for idx, network in enumerate(networks):
```
Done.

aabgd commented on src/python_testing/TC_CNET_4_11.py:
```diff
@@ -0,0 +1,611 @@
+
+        # 1. NetworkID is the hex representation of the ASCII values for PIXIT.CNET.WIFI_2ND_ACCESSPOINT_SSID
+        # 2. Connected is of type bool and is TRUE
+        for idx, network in enumerate(networks):
```
Done.

aabgd committed:
Merge branch 'master' into tc_cnet_4_11

aabgd committed:
Merge branch 'master' into tc_cnet_4_11

aabgd committed:
Merge branch 'master' into tc_cnet_4_11

aabgd committed:
Merge branch 'master' into tc_cnet_4_11

cecille commented on src/python_testing/TC_CNET_4_11.py:
```diff
@@ -0,0 +1,631 @@
+attr = cnet.Attributes
+
+
+async def connect_wifi_linux(ssid, password):
```
have you tested this on the test harness?

aabgd replied:
**Updated the script to run in the TH (Raspberry Pi + Docker)**

TH setup according to **"Matter Test-Harness User Manual"** v2.13+summer2025:
- 10.5 SDK Python Tests
- 10.5.1 Run Tests Inside SDK Docker Container

Configured Raspberry Pi with a LAN connection and connected through SSH.
Once logged in to Raspberry Pi:

Verify test TC_CNET_4_11.py is located in:
```
ubuntu@ubuntu:~/certification-tool/backend/test_collections/matter/sdk_tests/sdk_checkout/python_testing/scripts/sdk$
```

Run the following command from the Raspberry Pi terminal.
```
ubuntu@ubuntu:~$ cd ~/certification-tool
ubuntu@ubuntu:~$ ./backend/test_collections/matter/scripts/update-paa-certs.sh
```

Export environment variable:
```
cd ~/
ubuntu@ubuntu:~$ export PATH_TO_PAA_ROOTS="/var/paa-root-certs/"
```

Pull the Docker image:
```
cd ~/
ubuntu@ubuntu:~$ docker pull connectedhomeip/chip-cert-bins:a82e43e06e35c707f9016c38ee83712c2ab58966-patch
```

Create Docker container:
```
ubuntu@ubuntu:~$ docker run -v $PATH_TO_PAA_ROOTS:/paa_roots \
           -v /var/run/dbus/system_bus_socket:/var/run/dbus/system_bus_socket \
           -v /home/ubuntu/certification-tool/backend/test_collections/matter/sdk_tests/sdk_checkout/python_testing:/root/python_testing \
           -v $(pwd):/launch_dir \
           -v /sys/class/net:/sys/class/net \
           -v /dev:/dev \
           -v ~/test_logs:/root \
           --privileged --network host -it connectedhomeip/chip-cert-bins:a82e43e06e35c707f9016c38ee83712c2ab58966-patch /bin/bash
```

Once inside the container export environment variables:
```
root@ubuntu:~#
export DISCRIMINATOR="3840"
export SETUP_PASSCODE="20202021"
export WIFI_SSID1=<WIFI_SSID1>
export WIFI_PASSPHRASE1=<WIFI_PASSPHRASE1>
export WIFI_SSID2=<WIFI_SSID2>
export WIFI_PASSPHRASE2=<WIFI_PASSPHRASE2>
export BLE_INTERFACE_ID=0
```

Update and install necessary packages for managing WiFi:
```
root@ubuntu:~# apt-get update
root@ubuntu:~# apt-get install -y iw wpasupplicant iputils-ping sudo isc-dhcp-client
```

Create wpa_supplicant WiFi configuration file:
```
root@ubuntu:~# cat > /etc/wpa_supplicant/wpa_supplicant.conf <<EOF
ctrl_interface=/var/run/wpa_supplicant
update_config=1
country=ES

network={
    ssid="$WIFI_SSID1"
    psk="$WIFI_PASSPHRASE1"
    key_mgmt=WPA-PSK
    proto=RSN WPA
    pairwise=CCMP TKIP
    group=CCMP TKIP
    priority=1
}

network={
    ssid="$WIFI_SSID2"
    psk="$WIFI_PASSPHRASE2"
    key_mgmt=WPA-PSK
    proto=RSN WPA
    pairwise=CCMP TKIP
    group=CCMP TKIP
    priority=2
}
EOF
```

Change wpa_supplicant configuration file permissions:
```
root@ubuntu:~# chmod 600 /etc/wpa_supplicant/wpa_supplicant.conf
root@ubuntu:~# chown root:root /etc/wpa_supplicant/wpa_supplicant.conf
```

Start wpa_supplicant service and verify list of WiFi networks:
```
root@ubuntu:~# wpa_supplicant -B -i wlan0 -c /etc/wpa_supplicant/wpa_supplicant.conf -C /var/run/wpa_supplicant
```

May show the following error:
```
Failed to create interface p2p-dev-wlan0: -16 (Device or resource busy)
nl80211: Failed to create a P2P Device interface p2p-dev-wlan0
P2P: Failed to enable P2P Device interface
```
but this is a virtual interface, so do not worry about this one
Then verify both WiFi SSIDs are visible with:

```
root@ubuntu:~# wpa_cli list_networks
Selected interface 'wlan0'
network id / ssid / bssid / flags
0	MovistarFibra-B1FB50	any	
1	TP-Link_8434	any
```

From your computer, build and flash all-clusters-app to the ESP32:
```
idf.py -C examples/all-clusters-app/esp32 -p /dev/ttyUSB0 erase-flash
idf.py -C examples/all-clusters-app/esp32 -p /dev/ttyUSB0 flash monitor
```

Back in the Raspberry Pi terminal, run the Python Test:
```
root@ubuntu:~# python3 /root/python_testing/scripts/sdk/TC_CNET_4_11.py \
    --discriminator $DISCRIMINATOR \
    --passcode $SETUP_PASSCODE \
    --commissioning-method ble-wifi \
    --paa-trust-store-path /paa_roots \
    --storage-path admin_storage.json \
    --wifi-ssid $WIFI_SSID1 \
    --wifi-passphrase $WIFI_PASSPHRASE1 \
    --ble-interface-id $BLE_INTERFACE_ID \
    --string-arg PIXIT.CNET.WIFI_2ND_ACCESSPOINT_SSID:$WIFI_SSID2 \
    PIXIT.CNET.WIFI_2ND_ACCESSPOINT_CREDENTIALS:$WIFI_PASSPHRASE2 \
    --endpoint 0
```

If facing mDNS issues between runs, run the following command from the Raspberry Pi terminal to restart Avahi service:
```
ubuntu@ubuntu:~$ sudo systemctl restart avahi-daemon
ubuntu@ubuntu:~$ sudo systemctl status avahi-daemon
```

aabgd committed:
changed imports from chip.* to matter.* and ran isort

aabgd committed:
added teardown_class to reconnect to the original network and updated step 10

aabgd committed:
merging new connection script (sockets + wpa_supplicant) with test script

aabgd committed:
added run_command and changed connect_wifi_linux logic and added fallback to SSID1

aabgd committed:
resolved intermittent failures in steps 5 and 17

aabgd committed:
cleaned the code and tested on linux desktop

aabgd committed:
solved linter errors

aabgd committed:
testing in Raspberry Pi, changed wait periods

aabgd committed:
cleaned code, removed unnecessary logs and unused variables

aabgd committed:
also tested in Raspberry Pi 5 with Ubuntu Server

aabgd committed:
TC_CNET_4_11: Increase timeouts and improve robustness for Docker

- Increased NETWORK_CHANGE_TIMEOUT and MDNS_DISCOVERY_TIMEOUT to 180s for better reliability during network transitions and mDNS discovery.
- Improved subprocess handling to avoid log output interference.
- General robustness improvements for more consistent test execution.
- These changes are specifically aimed at ensuring reliable operation when running the test inside a Docker container.

aabgd committed:
TC_CNET_4_11: Comprehensive code optimization and refactoring

Major improvements to WiFi network switching test framework:

Code Organization:
- Extracted utility functions (find_network_and_assert, verify_operational_network) outside test class
- Separated concerns with dedicated helper functions

Route Management:
- Created centralized route management helpers (parse_default_routes, remove_lan_routes, etc.)
- Added temporarily_remove_lan_route_for_wifi for controlled network isolation

mDNS/Avahi Optimization:
- Created specialized mDNS cleanup functions with different intensities:
  * clean_avahi_cache(): Light cleanup for cache clearing
  * restart_avahi_daemon(): Medium cleanup with service restart
  * aggressive_mdns_cleanup(): Complete cleanup with all cached files
- Simplified setup_network_environment()
- Simplified cleanup_mdns_state()

This refactoring maintains full functionality while significantly improving
code organization, reducing duplication, and providing better error visibility.

aabgd committed:
refactored functions that used subprocess.run to use run_command function

aabgd committed:
Add mDNS discovery with avahi-browse for Docker environments
Implement progressive timeout reductions
Add session stability delays after network verification
Remove unnecessary network stabilization waits
Add comprehensive error handling and retry mechanisms

aabgd committed:
used avahi-browse for sensing mDNS

aabgd committed:
Restore TC_CNET_4_11.py to working version from f9b9342

aabgd committed:
Refactor TC_CNET_4_11: robust WiFi switching, LAN route removal in setup/teardown, improved subprocess handling

- Remove LAN default routes only in setup/teardown to force WiFi usage during test
- Refactor WiFi switching logic for reliability in both native Linux and Docker
- Use robust async subprocess execution for all shell/network commands
- Add stabilization waits and retries for network transitions
- Ensure all Matter commands and attribute checks are retried and validated

aabgd committed:
Major refactor: Add network robustness, mDNS discovery, and log filtering

Enhanced TC_CNET_4_11 with comprehensive network handling, async mDNS discovery
via zeroconf, DNS-SD log suppression (reduces 4500+ spam messages), improved
WiFi reliability with retry logic, and proper test environment management.

aabgd committed:
Refactor TC_CNET_4_11: function renames, helper extractions, and timeout optimizations

- Rename discover_dut_mdns → find_matter_devices_mdns for clarity
- Rename detect_wifi_iface → detect_wifi_interface for consistency
- Extract find_network_and_assert and verify_operational_network as standalone helper functions
- Optimize mDNS discovery timeout from 60s to 20s for faster execution
- Add detect_platform helper for better OS detection abstraction

aabgd committed:
Updated find_matter_devices_mdns() to use automatic device filtering using the device id

aabgd committed:
Merge branch 'master' into tc_cnet_4_11

aabgd committed:
removed unnecessary logs and corrected linter errors

aabgd committed:
restored connect_wifi_linux attempts logs

aabgd committed:
Merge branch 'master' into tc_cnet_4_11

aabgd committed:
Merge branch 'master' into tc_cnet_4_11

aabgd committed:
removed duplicated code from connect_wifi_linux

antonio-amjr reviewed (approved):
LGTM, approving.

From TH perspective, I believe the most important bit for me is presence of the fallback logic with the teardown method.
But just sharing a concern, I'm not sure it will be handled recovering from an error during execution though.

aabgd committed:
Merge branch 'master' into tc_cnet_4_11

aabgd committed:
Add contextlib import and fix remaining ruff errors

aabgd committed:
Fix indentation errors in TC_CNET_4_11.py

aabgd committed:
Add IPv6 support to mDNS discovery in TC_CNET_4_11

- Modified MatterServiceListener.add_service() to handle both IPv4 and IPv6 addresses
- Use socket.inet_ntoa() for IPv4 (4 bytes) and socket.inet_ntop() for IPv6 (16 bytes)
- Previously only IPv4 addresses were captured, causing intermittent discovery failures
- Matter devices may advertise IPv6 first, before obtaining IPv4 address
- This fixes race condition where device was not found if discovered before IPv4 assignment

aabgd committed:
Optimize TC_CNET_4_11: reduce MAX_ATTEMPTS to 4, adjust timeouts, improve network verification

aabgd committed:
TC_CNET_4_11: Remove redundant comments and fix ssid_str bug

- Remove ~30 redundant comments throughout the file
- Fix bug in connect_wifi_mac error messages using ssid instead of ssid_str
- Simplify verbose logging while keeping essential debug information
- Improve code readability without changing functionality

aabgd committed:
Fix RET505 linter error: Remove unnecessary else after return

aabgd committed:
updated waits and zeroconf

khodya committed:
make it rolling on TH

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

khodya committed:
fix ruff checks

khodya committed:
Merge branch 'master' into tc_cnet_4_11

zenghuchen-gg commented on src/python_testing/TC_CNET_4_11.py:
it would be great if the post-fix (time) could be unified, as I see some are using `_MS`, some are using `SECONDS`  and some are missing.

khodya replied:
Good catch. Fixed! Thanks

khodya committed:
spec alignments

khodya committed:
adjust timing constants names

cecille commented:
Summary - we discussed this in a meeting with Karsten last week. The HRAP team has some designs for starting up connectable SSIDs or various types from within the test harness. What we discussed in the meeting was a two phase approach. 1) We land this as-is because it matches the current setup used by the ATLs with two external Wi-Fis. 2) The HRAP team implements the TH changes they need to support the TH supplying the Wi-Fi through a secondary interface (required for PDC). The HRAP team will lead that effort with support from TH team for logistics (identification of hardware, discussion with ATLs etc.). 3) Once the HRAP team is through SVE and the network spin-up is stable, our team will go back and make the CNET tests easier to use.

cecille commented:
@ksperling-apple - are you OK if we land this test as-is for 1.7?

cecille reviewed (approved):
This one has been up for a long time, tested by multiple people. Doesn't look like there are more reviews incoming, labeling.
